enterprise/fulltext-addon/src/main/java/org/neo4j/kernel/api/impl/fulltext/ScoreEntityIterator.java

```diff
@@ -105,35 +108,95 @@ public static ScoreEntityIterator emptyIterator()
 
     private static class ConcatenatingScoreEntityIterator extends ScoreEntityIterator
     {
-        private final Iterator<? extends ScoreEntityIterator> iterators;
+        private final List<? extends ScoreEntityIterator> iterators;
+        private final ScoreEntry[] buffer;
         private ScoreEntityIterator currentIterator;
+        private boolean fetched;
+        private ScoreEntry nextHead;
 
-        ConcatenatingScoreEntityIterator( Iterator<? extends ScoreEntityIterator> iterators )
+        ConcatenatingScoreEntityIterator( List<? extends ScoreEntityIterator> iterators )
         {
             this.iterators = iterators;
+            this.buffer = new ScoreEntry[iterators.size()];
         }
 
         @Override
         public boolean hasNext()
         {
-            if ( currentIterator == null || !currentIterator.hasNext() )
+            if ( !fetched )
+            {
+                fetch();
+            }
+            return nextHead != null;
+        }
+
+        private void fetch()
+        {
+            int candidateHead = -1;
+            for ( int i = 0; i < iterators.size(); i++ )
```

Comment:
do not think its that important in this case since we will have one iterator (ok, ok - two) in mostly in all the cases, but if in future or you still think it's important to make this part faster we can use binary heap here instead.